### PR TITLE
Implement `conflicts_with` stanza

### LINKED ISSUE #4497: add to DSL: conflicts_with :formula
as discussed in #4451 .

### PR DESCRIPTION
**Pull request by jawshooah**
Closes #4497.

As mentioned in the commit message for 46738c9, most of `conflicts_with.rb` was taken directly from `depends_on.rb` and modified accordingly. It would be likely be better to outsource common methods to a shared superclass or utility class.

### CODE REVIEW

File: lib/hbc/dsl/conflicts_with.rb [hunk 1]
```diff
@@ -9,18 +9,133 @@ class Hbc::DSL::ConflictsWith
...
+  VALID_ARCHES = [
+                  # category
+                  :intel,
+                  :ppc,
+                  # specific
+                  :i386,
+                  :x86_64,
+                  :ppc_7400,
+                  :ppc_64,
+                 ]
```
Comment (phinze) [inline lib/hbc/dsl/conflicts_with.rb]: Do we support machines running PPC?

File: lib/hbc/dsl/conflicts_with.rb [hunk 2]
```diff
@@ -9,18 +9,133 @@ class Hbc::DSL::ConflictsWith
...
+                   :intel32    => :i386,
+                   :intel_64   => :x86_64,
+                   :intel64    => :x86_64,
+                   :amd_64     => :x86_64,
+                   :amd64      => :x86_64,
+                   :ppc7400    => :ppc_7400,
+                   :ppc_32     => :ppc_7400,
+                   :ppc32      => :ppc_7400,
+                   :ppc64      => :ppc_64,
+                  }
```
Comment (phinze) [inline lib/hbc/dsl/conflicts_with.rb]: I like the thought behind this is nice for UX, but in this case I think it's more important to reduce our surface area of possible Cask content than to allow for spelling variants.

File: lib/hbc/dsl.rb [hunk 3]
```diff
@@ -201,16 +201,14 @@ def depends_on(*args)
...
 
     def conflicts_with(*args)
-      if @conflicts_with and !args.empty?
-        # todo: remove this constraint, and instead merge multiple conflicts_with stanzas
-        raise Hbc::CaskInvalidError.new(self.token, "'conflicts_with' stanza may only appear once")
-      end
       return @conflicts_with if args.empty?
-      @conflicts_with ||= begin
-        Hbc::DSL::ConflictsWith.new(*args) unless args.empty?
+      @conflicts_with ||= Hbc::DSL::ConflictsWith.new()
```
Comment (phinze) [inline lib/hbc/dsl.rb]: Ruby style nitpick - empty parents are dropped in method calls, so -> `ConflictsWith.new`

File: lib/hbc/dsl/conflicts_with.rb [hunk 4]
```diff
@@ -9,18 +9,133 @@ class Hbc::DSL::ConflictsWith
...
+
   attr_accessor *VALID_KEYS
```
Comment (phinze) [inline lib/hbc/dsl/conflicts_with.rb]: The `attr_accessor` directive defines simple reader and writer methods, which you then override below. If you change this to `attr_reader` you can drop all the simple readers below.

File: lib/hbc/dsl/conflicts_with.rb [hunk 5]
```diff
@@ -9,18 +9,133 @@ class Hbc::DSL::ConflictsWith
...
+        raise
+      end
+    rescue StandardError => e
+      raise "invalid 'conflicts_with :macos' value: #{arg.inspect}"
     end
   end
 
+  def formula
+    @formula
+  end
```
Comment (phinze) [inline lib/hbc/dsl/conflicts_with.rb]: Like this one! See `attr_reader` note above.

File: lib/hbc/installer.rb [hunk 6]
```diff
@@ -206,6 +207,77 @@ def cask_dependencies
...
+      end
+    elsif @cask.conflicts_with.macos.length > 1
+      if @cask.conflicts_with.macos.include?(Gem::Version.new(MacOS.release.to_s))
+        raise Hbc::CaskError.new "Cask #{@cask} conflicts with OS X release being one of: #{@cask.conflicts_with.macos(&:to_s).inspect}, and you are running release #{MacOS.release}."
+      end
+    else
+      if MacOS.release == @cask.conflicts_with.macos.first
+        raise Hbc::CaskError.new "Cask #{@cask} conflicts with OS X release #{@cask.conflicts_with.macos.first}, and you are running release #{MacOS.release}."
+      end
+    end
```
Comment (phinze) [inline lib/hbc/installer.rb]: It's sorta finicky to have to switch on type here.

What about pulling all this up into an object you can ask? Something like `MacOsConflict` and then up here you could do a nice little

```ruby
def macos_conflicts
   Array(@cask.conflicts.macos).each do |c|
     raise Hbc::CaskError.new("Wrong OS! Needed: #{c} Have: #{MacOS.release}") unless c.satisfied_by?(MacOS.release)
   end
end
```

File: lib/hbc/installer.rb [hunk 7]
```diff
@@ -206,6 +207,77 @@ def cask_dependencies
...
+
+  def arch_conflicts
+    return unless @cask.conflicts_with.arch
+    @current_arch ||= [
+                       Hardware::CPU.type,
+                       Hardware::CPU.is_32_bit? ?
+                         (Hardware::CPU.intel? ? :i386   : :ppc_7400) :
+                         (Hardware::CPU.intel? ? :x86_64 : :ppc_64)
+                      ]
+    return if Array(@cask.conflicts_with.arch & @current_arch).empty?
```
Comment (phinze) [inline lib/hbc/installer.rb]: Strong `&` set intersection work here! :grinning:

File: lib/hbc/installer.rb [hunk 8]
```diff
@@ -206,6 +207,77 @@ def cask_dependencies
...
+  def x11_conflicts
+    return unless @cask.conflicts_with.x11
+    unless Hbc.x11_libpng.select(&:exist?).empty?
+      raise Hbc::CaskError.new "Cask #{@cask} conflicts with XQuartz/X11, and you have it installed."
+    end
+  end
+
+  def formula_conflicts
+    return unless @cask.conflicts_with.formula and not @cask.conflicts_with.formula.empty?
+    @cask.conflicts_with.formula.each do |dep_name|
```
Comment (phinze) [inline lib/hbc/installer.rb]: Again just coerce with `Array()` and you can drop the guard.

File: lib/hbc/installer.rb [hunk 9]
```diff
@@ -206,6 +207,77 @@ def cask_dependencies
...
+  def satisfy_conflicts
+    if @cask.conflicts_with
+      ohai 'Satisfying conflicts'
+      macos_conflicts
+      arch_conflicts
+      x11_conflicts
+      formula_conflicts
+      cask_conflicts
+      puts 'complete'
+    end
```
Comment (phinze) [inline lib/hbc/installer.rb]: I think you noted this in a test comment, but this is begging for an object and an interface.

Something like:

```ruby
ohai 'Satisfiying conflicts'
@cask.conflicts_with.each each do |conflicts_with|
  conflicts_with.check!
end
puts 'complete'

**Comment by phinze** [comment]
Nice work on this so far @jawshooah - thank you! Left you plenty of review notes - happy to clarify and discuss any of 'em. :+1:

File: lib/hbc/dsl/conflicts_with.rb (hunk 2 above)
Comment (jawshooah) [inline lib/hbc/dsl/conflicts_with.rb]: I'm inclined to agree; like most of the changes in `conflicts_with.rb`, this was pulled directly from `depends_on.rb`, and I didn't want to remove _existing_ functionality.

**Comment by jawshooah** [comment]
Thanks for the speedy response! I love a good code review. 

As noted in the opening comment, most of the changes here (including much of what you've commented on :wink:) were taken directly from `depends_on`. When addressing your suggestions, should I make similar changes to that implementation as well?

File: lib/hbc/dsl/conflicts_with.rb (hunk 1 above)
Comment (jawshooah) [inline lib/hbc/dsl/conflicts_with.rb]: [Seems we do](https://github.com/caskroom/homebrew-cask/pull/8085).

**Comment by phinze** [comment]
> When addressing your suggestions, should I make similar changes to that implementation as well?

If you're willing, that'd be great! Though I'd probably approach that as a separate "refactor depends_on" PR to keep things from getting too hefty.

**Comment by rolandwalker** [comment]
I recommend against this degree of duplication of code.

**Comment by jawshooah** [comment]
@rolandwalker Agreed, I'm just not entirely sure of the best way to go about deduplicating it. It seems like every class in `Hbc::DSL` is intended to map one-to-one with a DSL stanza, so I don't know if creating a `Hbc::DSL::Dependency` or similar superclass would be a breach of style or expectations. What do you think?

**Comment by jawshooah** [comment]
One obvious thing that would help would be to move `self.coerce_os_release` to `Hbc::MacOS`. #9329

**Comment by phinze** [comment]
Ah yeah I didn't see the amount of similarity this had with depends_on in my initial review. I think `Hbc::DSL::Dependency` as either a superclass or a mixin is a great idea.

**Comment by jawshooah** [comment]
@phinze I've moved all shared logic to `Hbc::DSL::Dependency` and `Hbc::Dependencies`.

The `depends_on` refactor is all in 6503dbc (build failure seems unrelated), so if desired I can leave that out and save it for another PR.

**Comment by jawshooah** [comment]
Turns out it wasn't so unrelated; the `pairs` method of `DependsOn` wasn't stubbed out in the tests that failed, and they were skipped on my machine. Pushed a fix, all tests should pass now.

**Comment by rolandwalker** [comment]
Procedurally:

While we are pretty darn good about writing roadmaps in this project, it is not practical to get 100% of our private branches, thoughts, and IRC discussions into GitHub issues.  So, there are some things about implementing `conflicts_with` that haven't been outlined, about which, see below.

When the idea of implementing `conflicts_with` was mentioned in #9202, I responded by self-assigning #4497.  As explained in https://github.com/caskroom/homebrew-cask/issues/8784#issuecomment-69584246, that was a quick way of saying "Heads-up guys – working on this thingy is likely to involve wasted effort".

Substantively:

* `conflicts_with` and `depends_on` are conceptually and architecturally related.  The features are being built out in a certain order, and `conflicts_with` was intentionally deferred.
* `depends_on` was fully separated from the old `depends_on_formula` in #7401 so that we could make it general.
* Improvements and support for various new keys were added in a series of subsequent PRs (*eg* `depends_on :macos` in #7696) – the easy cases.
* In #8491 and others, @ndr-qef made fixes, and added the harder case of `depends on :cask`, introducing a proper dependency graph.

Roughly, the path leading to `conflicts_with` was:
* factor dependency checking out of `installer.rb`
* generalize the dependency graph to include all `depends_on` keys (not just `:cask`), so that we can walk the graph and verify that all dependencies are installable before acting
* implement `conflicts_with` in terms of the dependency graph

Currently, the folks interested in dependencies are discussing whether to add interface to support pure meta-Casks (#8492).  It is worth noting that there is a divergence of opinions about whether or not to support meta-Casks, and what the interface might be.  The maintainers are responding to that by taking their time, being consultative, going to extra effort to collect opposite views.  It's a tautology, but: the only way we can work together is by coordinating, asking questions, and listening.

Scanning this PR, conceptual flaws are apparent, such as the notion that one can [serially](https://github.com/jawshooah/homebrew-cask/blob/0633925390e925b398e281104e133d3849e2a570/lib/hbc/installer.rb#L63) `satisfy_dependencies`, `satisfy_conflicts`, or [flatten/ignore the dependency graph](https://github.com/jawshooah/homebrew-cask/blob/0633925390e925b398e281104e133d3849e2a570/lib/hbc/dependencies.rb#L163) in "conflict mode".  Nor are impossible `depends_on`/`conflicts_with` combinations detected.

Furthermore, it is too large.  `depends_on` was developed over [at least a dozen PRs](https://github.com/caskroom/homebrew-cask/search?q=depends_on&type=Issues&utf8=%E2%9C%93) with ample space for discussion.

**Comment by phinze** [comment]
Ahhhh ok! Thanks for laying out the full context so clearly @rolandwalker. I was running behind on the overall conversation here, and I dove into line-by-line review before I should have. Sorry to steer you in the wrong direction @jawshooah!

Sounds like we're going to need to back up a little bit on this feature: resolve the direction in @rolandwalker's linked conversations, then break down the pieces so we can avoid any big-bang PRs.

Thanks for diving into this @jawshooah - once I'm a little more caught back up I promise I won't lead you waste work like this. :confused: We may not merge this work as is, but it definitely helps to inform the progression of the feature as a whole, so thanks for that! :+1:

**Comment by jawshooah** [comment]
Thanks to @rolandwalker for context, clarification, and critique. @phinze, don't worry about it. I should have asked around more before diving headfirst into the implementation.

Is there currently an open discussion around implementing a more generalized dependency graph? It seems like resolution would proceed in the following order:

* Generate a directed graph, where:
  * Each node represents a `:cask`, `:formula`, `:macos`, `:arch`, `:x11`, or `:java` definition (with the potential for new types to be added later on)
  * Each directed edge represents a relationship from the source to the target, either a conflict or a dependency.
* Check that the graph is acyclic. If not, then either:
  * We have a cyclic dependency, or
  * At least one dependency conflicts with another.

Obviously this is going to take some deeper thinking, but that's the gist as I understand it.

We will need to interface with Homebrew to get `formula` dependencies and conflicts, the latter being inaccessible through the `brew` CLI, and the former limited to `formula -> formula` dependencies. We may need to either restore some of `homebrew-fork` or write our own `formula` parser/loader to get the rest.

We'll also need to hardcode in relationships between `arch`, `macos`, `x11`, and `java`.

This is going to be quite an endeavor.

**Comment by rolandwalker** [comment]
> We will need to interface with Homebrew to get formula dependencies and conflicts,

Very nice catch!  I intentionally deleted that bullet point, in order not to muddy the waters, and added the weasel-word "roughly" to cover my behind.  It is probably best not to try to resolve Formula dependencies, and just pretend that it works until we encounter breakage.  It's too much work!

`depends_on` is not fully born yet.  Bringing it the rest of the way into the world is not a terribly forbidding endeavor.  Or shouldn't be.

And then, if `depends_on` is expressed well in code, `conflicts_with` should fall out fairly naturally as an inverse-dependency.  There are other package managers whose code we could consult if for some reason we got stuck.

Almost nothing is "expressed well" at the moment, because of the transition phase of #4688.  In large part, git history would show that I took @phinze's designs and subjected them to … bending and denting.

It's a design lesson!  Good design will survive the next guy.

**Comment by jawshooah** [comment]
Ignoring Formula dependencies makes me feel queasy, but it probably won't cause any problems in practice, and I agree that the effort required to handle them properly is probably too much, at least for an initial implementation.

Though this first attempt was a bust, I'm still very eager to contribute! Is there any way that I can help to bring proper dependency resolution to fruition? I don't want to step on any toes, but there must be something I can hack away at.

I care a lot about this project, and I'm happy to help however I can :bowtie:

**Comment by jawshooah** [comment]
@rolandwalker Is anyone currently working on a comprehensive solution for dependency resolution as you described? I wasn't sure whether your response to this PR was an invitation to give it another, better-conceived shot or a polite request to let the big boys handle this one :wink:

**Pull request closed without merging**